In [64]:
import pandas as pd

file_path = "../data/raw/2019-Nov.csv"

df = pd.read_csv(file_path, nrows=100_000)

df.head()

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
0,2019-11-01 00:00:00 UTC,view,1003461,2053013555631882655,electronics.smartphone,xiaomi,489.07,520088904,4d3b30da-a5e4-49df-b1a8-ba5943f1dd33
1,2019-11-01 00:00:00 UTC,view,5000088,2053013566100866035,appliances.sewing_machine,janome,293.65,530496790,8e5f4f83-366c-4f70-860e-ca7417414283
2,2019-11-01 00:00:01 UTC,view,17302664,2053013553853497655,NaN,creed,28.31,561587266,755422e7-9040-477b-9bd2-6a6e8fd97387
3,2019-11-01 00:00:01 UTC,view,3601530,2053013563810775923,appliances.kitchen.washer,lg,712.87,518085591,3bfb58cd-7892-48cc-8020-2f17e6de6e7f
4,2019-11-01 00:00:01 UTC,view,1004775,2053013555631882655,electronics.smartphone,xiaomi,183.27,558856683,313628f1-68b8-460d-84f6-cec7a8796ef2


In [65]:
# data information
# one row = one user event involving one product at one in time

df.shape # 9 columns and 100k rows

(100000, 9)

In [66]:
df.head()

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
0,2019-11-01 00:00:00 UTC,view,1003461,2053013555631882655,electronics.smartphone,xiaomi,489.07,520088904,4d3b30da-a5e4-49df-b1a8-ba5943f1dd33
1,2019-11-01 00:00:00 UTC,view,5000088,2053013566100866035,appliances.sewing_machine,janome,293.65,530496790,8e5f4f83-366c-4f70-860e-ca7417414283
2,2019-11-01 00:00:01 UTC,view,17302664,2053013553853497655,NaN,creed,28.31,561587266,755422e7-9040-477b-9bd2-6a6e8fd97387
3,2019-11-01 00:00:01 UTC,view,3601530,2053013563810775923,appliances.kitchen.washer,lg,712.87,518085591,3bfb58cd-7892-48cc-8020-2f17e6de6e7f
4,2019-11-01 00:00:01 UTC,view,1004775,2053013555631882655,electronics.smartphone,xiaomi,183.27,558856683,313628f1-68b8-460d-84f6-cec7a8796ef2


In [67]:
df.tail()

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
99995,2019-11-01 03:52:45 UTC,view,1004781,2053013555631882655,electronics.smartphone,huawei,254.71,514148475,03f3fa00-1c0a-48e3-9cd1-f9ae4b11809e
99996,2019-11-01 03:52:45 UTC,view,2600128,2053013563970159485,NaN,NaN,175.55,560904986,344e7ccb-169c-48bc-8b9b-b93bce85391e
99997,2019-11-01 03:52:45 UTC,view,1005159,2053013555631882655,electronics.smartphone,xiaomi,212.10,566309730,72b2106a-c885-41d5-99ab-bbed9fa730f7
99998,2019-11-01 03:52:45 UTC,cart,1801906,2053013554415534427,electronics.video.tv,tcl,257.33,564331418,092b3c00-2d01-40e7-a1d9-dfa9bbad847e
99999,2019-11-01 03:52:45 UTC,view,3100954,2053013555262783879,appliances.kitchen.blender,scarlett,47.59,536691296,e8c0cad6-58a7-46e9-8fcd-3ff49f37fb46


In [68]:
df.sample(10)

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
1088,2019-11-01 00:07:30 UTC,view,1005105,2053013555631882655,electronics.smartphone,apple,1348.61,540350609,c7f362e3-d08e-4eaa-8a34-a07b1966b128
34285,2019-11-01 02:20:15 UTC,view,1801905,2053013554415534427,electronics.video.tv,samsung,1029.37,518435836,e3a7d3ee-b2d0-44a0-b6ba-4ebb206ef69c
75132,2019-11-01 03:23:49 UTC,view,7600004,2053013552821698803,NaN,tp-link,17.94,549734298,94a928ba-3b3a-4feb-a75c-e00867a5b89b
35810,2019-11-01 02:23:16 UTC,view,15600023,2053013559767466645,NaN,NaN,353.93,513197327,fc45ac44-39f3-4aa1-9bff-28685bd06057
66414,2019-11-01 03:12:31 UTC,view,2501888,2053013564003713919,appliances.kitchen.oven,artel,41.77,561800588,ce1d03b3-e8bb-4860-a3f0-8ee4c6c03f49
28853,2019-11-01 02:08:49 UTC,view,1801690,2053013554415534427,electronics.video.tv,samsung,369.45,555967953,bb03a4f0-e74c-4bfa-9d21-64a231b93eff
69765,2019-11-01 03:17:03 UTC,view,1801559,2053013554415534427,electronics.video.tv,philips,321.73,514415218,0036a270-5f74-4689-81f9-230c3da82c20
51874,2019-11-01 02:51:30 UTC,view,31501166,2053013558031024687,NaN,carlsbad,281.04,536860813,0d35bc9f-0e98-447a-a4ec-423164a7b013
86019,2019-11-01 03:36:52 UTC,view,1004741,2053013555631882655,electronics.smartphone,xiaomi,189.97,527681959,6d1e1022-fc50-43c3-9473-cd1e76a64eef
98732,2019-11-01 03:51:27 UTC,view,12702457,2053013553559896355,NaN,matador,57.92,522560074,16fa6f9e-bc61-4742-818b-a52dd652619e


In [69]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   event_time     100000 non-null  str    
 1   event_type     100000 non-null  str    
 2   product_id     100000 non-null  int64  
 3   category_id    100000 non-null  int64  
 4   category_code  66160 non-null   str    
 5   brand          84224 non-null   str    
 6   price          100000 non-null  float64
 7   user_id        100000 non-null  int64  
 8   user_session   100000 non-null  str    
dtypes: float64(1), int64(3), str(5)
memory usage: 6.9 MB


In [70]:
df.describe(include="all")

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
count,100000,100000,1.000000e+05,1.000000e+05,66160,84224,100000.000000,1.000000e+05,100000
unique,12925,3,NaN,NaN,121,1594,NaN,NaN,24889
top,2019-11-01 03:31:06 UTC,view,NaN,NaN,electronics.smartphone,samsung,NaN,NaN,0c239db5-422b-4198-b76a-035ffcc275fa
freq,30,97489,NaN,NaN,24928,11748,NaN,NaN,140
mean,NaN,NaN,1.123577e+07,2.057926e+18,NaN,NaN,281.539756,5.354871e+08,NaN
std,NaN,NaN,1.219807e+07,1.965420e+16,NaN,NaN,342.414641,2.031755e+07,NaN
min,NaN,NaN,1.000978e+06,2.053014e+18,NaN,NaN,0.000000,2.752567e+08,NaN
25%,NaN,NaN,1.200954e+06,2.053014e+18,NaN,NaN,65.640000,5.159248e+08,NaN
50%,NaN,NaN,5.301584e+06,2.053014e+18,NaN,NaN,164.950000,5.315900e+08,NaN
75%,NaN,NaN,1.650001e+07,2.053014e+18,NaN,NaN,350.852500,5.549301e+08,NaN


### Data Type Observations

The initial schema inspection identified several columns whose loaded data types
do not fully represent their analytical meaning:

- `event_time` is stored as a string but represents a timestamp.
- `product_id` is stored as an integer but represents a product identifier.
- `category_id` is stored as an integer but represents a category identifier.
- `user_id` is stored as an integer but represents a user identifier.

The identifier columns should not be treated as quantitative variables because
operations such as mean, standard deviation, and quartiles have no analytical
meaning for IDs.

In [71]:
df["event_time"] = pd.to_datetime(df["event_time"], utc=True)
df["event_time"].dtype

datetime64[us, UTC]

In [72]:
id_columns = ["product_id", "category_id", "user_id"]

df[id_columns] = df[id_columns].astype("string")

df[id_columns].dtypes


product_id     string
category_id    string
user_id        string
dtype: object

In [73]:
# updated dataset with correct data types

df.describe(include="all")


,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
count,100000,100000,100000,100000,66160,84224,100000.000000,100000,100000
unique,NaN,3,20675,557,121,1594,NaN,20778,24889
top,NaN,view,1004856,2053013555631882655,electronics.smartphone,samsung,NaN,513002539,0c239db5-422b-4198-b76a-035ffcc275fa
freq,NaN,97489,1132,24928,24928,11748,NaN,140,140
mean,2019-11-01 02:35:10.685879+00:00,NaN,NaN,NaN,NaN,NaN,281.539756,NaN,NaN
min,2019-11-01 00:00:00+00:00,NaN,NaN,NaN,NaN,NaN,0.000000,NaN,NaN
25%,2019-11-01 02:00:17+00:00,NaN,NaN,NaN,NaN,NaN,65.640000,NaN,NaN
50%,2019-11-01 02:48:26.500000+00:00,NaN,NaN,NaN,NaN,NaN,164.950000,NaN,NaN
75%,2019-11-01 03:23:39+00:00,NaN,NaN,NaN,NaN,NaN,350.852500,NaN,NaN
max,2019-11-01 03:52:45+00:00,NaN,NaN,NaN,NaN,NaN,2574.070000,NaN,NaN


In [74]:
summary = pd.DataFrame({
    "dtype": df.dtypes,
    "missing": df.isna().sum(),
    "missing_pct": round(df.isna().mean() * 100, 2),
    "unique": df.nunique()
})

summary

,dtype,missing,missing_pct,unique
event_time,"datetime64[us, UTC]",0,0.00,12925
event_type,str,0,0.00,3
product_id,string,0,0.00,20675
category_id,string,0,0.00,557
category_code,str,33840,33.84,121
brand,str,15776,15.78,1594
price,float64,0,0.00,8069
user_id,string,0,0.00,20778
user_session,str,0,0.00,24889


In [75]:
df.loc[df["category_code"].isna(), "category_id"].unique()

<StringArray>
['2053013553853497655', '2053013558190408249', '2053013553559896355',
 '2060981320581906480', '2053013562183385881', '2053013563391345499',
 '2053013553375346967', '2053013562837697343', '2053013558752445019',
 '2090228413959766319',
 ...
 '2053013563273904979', '2059484387635888645', '2053013556403634639',
 '2053013565589160919', '2053013559222207091', '2141355067922449282',
 '2104564977263182827', '2071489994693804498', '2053013557737423389',
 '2053013557695480347']
Length: 334, dtype: string

In [76]:
ids_with_missing = df.loc[df["category_code"].isna(), "category_id"].unique()

matches = df[
    df["category_id"].isin(ids_with_missing)
    & df["category_code"].notna()
]

matches[["category_id", "category_code"]].drop_duplicates()

,category_id,category_code


In [77]:
df["event_type"].value_counts()

event_type
view        97489
purchase     1422
cart         1089
Name: count, dtype: int64

In [78]:
df["event_type"].value_counts(normalize=True)

event_type
view        0.97489
purchase    0.01422
cart        0.01089
Name: proportion, dtype: float64

In [79]:
df["price"].describe().round(2)

count    100000.00
mean        281.54
std         342.41
min           0.00
25%          65.64
50%         164.95
75%         350.85
max        2574.07
Name: price, dtype: float64

## price

The average cost of items are $281.54
the cheapest item is $0
the most expensive product is 2574.07

In [80]:
df.nsmallest(20, "price")


,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
6258,2019-11-01 00:38:01+00:00,view,33100000,2058719826188173878,NaN,NaN,0.0,546996930,969ea68f-a919-4d32-8925-b709d87c539c
7245,2019-11-01 00:42:51+00:00,view,33100000,2058719826188173878,NaN,NaN,0.0,546996930,b1ab3863-bbf5-4370-a3d8-bfa70580dd09
12743,2019-11-01 01:07:15+00:00,view,12720812,2053013553559896355,NaN,NaN,0.0,516269492,9bf68f2a-fd78-4b19-97ac-7d838c82eb4c
12908,2019-11-01 01:07:58+00:00,view,12720812,2053013553559896355,NaN,NaN,0.0,516269492,9bf68f2a-fd78-4b19-97ac-7d838c82eb4c
13503,2019-11-01 01:26:19+00:00,view,38900075,2085718636156158307,NaN,NaN,0.0,539587206,4d91487b-2b8d-41e3-9328-f8ab325a95cf
17220,2019-11-01 01:40:17+00:00,view,13902536,2053013557343158789,construction.components.faucet,NaN,0.0,514119888,793320f2-aab3-4696-b249-94908ec73971
18246,2019-11-01 01:43:13+00:00,view,2702683,2053013563911439225,appliances.kitchen.refrigerators,NaN,0.0,552636285,ed87f60a-412c-4d28-8a21-c0d261619c60
19682,2019-11-01 01:46:58+00:00,view,1307551,2053013558920217191,computers.notebook,NaN,0.0,518736272,bc56ca44-6875-4b55-8861-a5c2cd48fd75
20363,2019-11-01 01:48:48+00:00,view,4300485,2053013552385491165,NaN,NaN,0.0,530173538,600b6494-b8b3-455c-9694-6d72556cb1c2
20418,2019-11-01 01:48:57+00:00,view,4300485,2053013552385491165,NaN,NaN,0.0,530173538,600b6494-b8b3-455c-9694-6d72556cb1c2


In [81]:
df.nlargest(20, "price")

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
1086,2019-11-01 00:07:29+00:00,view,21408165,2053013561579406073,electronics.clocks,NaN,2574.07,513118352,4c14bf2a-2820-4504-929d-046356a5a204
1167,2019-11-01 00:08:00+00:00,view,21408165,2053013561579406073,electronics.clocks,NaN,2574.07,513118352,4c14bf2a-2820-4504-929d-046356a5a204
7728,2019-11-01 00:45:07+00:00,view,21408160,2053013561579406073,electronics.clocks,NaN,2574.07,512376038,3ec14458-4d55-4f75-adc3-72b0ad9ad4bb
7864,2019-11-01 00:45:48+00:00,view,21408160,2053013561579406073,electronics.clocks,NaN,2574.07,512376038,3ec14458-4d55-4f75-adc3-72b0ad9ad4bb
33302,2019-11-01 02:18:16+00:00,view,21408156,2053013561579406073,electronics.clocks,NaN,2574.07,515129276,dacc1c2b-d20e-4faa-9f9f-045389ddbcf6
36825,2019-11-01 02:25:15+00:00,view,21408154,2053013561579406073,electronics.clocks,NaN,2574.07,530545378,38eb2a01-c7c4-411d-abd4-f8b659940ebe
37007,2019-11-01 02:25:36+00:00,view,21408154,2053013561579406073,electronics.clocks,NaN,2574.07,530545378,38eb2a01-c7c4-411d-abd4-f8b659940ebe
37219,2019-11-01 02:26:03+00:00,view,21408165,2053013561579406073,electronics.clocks,NaN,2574.07,530545378,38eb2a01-c7c4-411d-abd4-f8b659940ebe
37321,2019-11-01 02:26:16+00:00,view,21408165,2053013561579406073,electronics.clocks,NaN,2574.07,530545378,38eb2a01-c7c4-411d-abd4-f8b659940ebe
66626,2019-11-01 03:12:48+00:00,view,21407288,2053013561579406073,electronics.clocks,rado,2574.07,538669191,e7f63535-a8b9-4c10-b51b-a8d9166e81a7


the most expensive product that costs $2574.07 is electinics clocks.
for those cost $0, we need more investigation. 

# Time

In [82]:
df["event_time"].min()


Timestamp('2019-11-01 00:00:00+0000', tz='UTC')

In [83]:
df["event_time"].max()

Timestamp('2019-11-01 03:52:45+0000', tz='UTC')

the data only contains November 1st, 2029.

# entity

In [84]:
print("unique user_id:", df["user_id"].nunique())
print("unique user_session:", df["user_session"].nunique())
print("unique product_id:", df["product_id"].nunique())
print("unique category_id:", df["category_id"].nunique())
print("unique brand:", df["brand"].nunique())

unique user_id: 20778
unique user_session: 24889
unique product_id: 20675
unique category_id: 557
unique brand: 1594


In [85]:
entities = {
    "events": len(df),
    "users": df["user_id"].nunique(),
    "sessions": df["user_session"].nunique(),
    "products": df["product_id"].nunique(),
    "categories": df["category_id"].nunique(),
    "brands": df["brand"].nunique()
}

entities

{'events': 100000,
 'users': 20778,
 'sessions': 24889,
 'products': 20675,
 'categories': 557,
 'brands': 1594}

The dataset contains 100k events generated by 20778 users across 24889 sessions.

In [86]:
df.duplicated().sum()

np.int64(27)

In [ ]:
# check duplicates
duplicates = df[df.duplicated(keep=False)]

duplicates.sort_values(
    ["user_id", "user_session", "event_time"]
).head(30)

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
16136,2019-11-01 01:36:58+00:00,view,1004173,2053013555631882655,electronics.smartphone,xiaomi,164.36,512595904,85f11e30-0d57-452d-aadc-bd110497aa7c
16138,2019-11-01 01:36:58+00:00,view,1004173,2053013555631882655,electronics.smartphone,xiaomi,164.36,512595904,85f11e30-0d57-452d-aadc-bd110497aa7c
99539,2019-11-01 03:52:17+00:00,view,7600143,2053013552821698803,NaN,tenda,50.12,514677746,08a0f1dc-17bd-479e-b977-c237f7b7a879
99541,2019-11-01 03:52:17+00:00,view,7600143,2053013552821698803,NaN,tenda,50.12,514677746,08a0f1dc-17bd-479e-b977-c237f7b7a879
44967,2019-11-01 02:39:56+00:00,cart,30000048,2127425436764865054,construction.tools.welding,NaN,118.90,515136785,d03d2aa1-c23f-4b0d-bdf9-4d9dffc0407d
44968,2019-11-01 02:39:56+00:00,cart,30000048,2127425436764865054,construction.tools.welding,NaN,118.90,515136785,d03d2aa1-c23f-4b0d-bdf9-4d9dffc0407d
44969,2019-11-01 02:39:56+00:00,cart,30000048,2127425436764865054,construction.tools.welding,NaN,118.90,515136785,d03d2aa1-c23f-4b0d-bdf9-4d9dffc0407d
18998,2019-11-01 01:45:15+00:00,cart,1305215,2053013558920217191,computers.notebook,apple,1209.55,515742799,ca33e1b9-ecf5-4b50-ba76-007248bad43d
19000,2019-11-01 01:45:15+00:00,cart,1305215,2053013558920217191,computers.notebook,apple,1209.55,515742799,ca33e1b9-ecf5-4b50-ba76-007248bad43d
94113,2019-11-01 03:46:16+00:00,cart,1004833,2053013555631882655,electronics.smartphone,samsung,172.23,516205526,b528bae4-f776-4720-801f-8ee1ac565ad2


In [88]:
df.drop_duplicates()

,event_time,event_type,product_id,category_id,category_code,brand,price,user_id,user_session
0,2019-11-01 00:00:00+00:00,view,1003461,2053013555631882655,electronics.smartphone,xiaomi,489.07,520088904,4d3b30da-a5e4-49df-b1a8-ba5943f1dd33
1,2019-11-01 00:00:00+00:00,view,5000088,2053013566100866035,appliances.sewing_machine,janome,293.65,530496790,8e5f4f83-366c-4f70-860e-ca7417414283
2,2019-11-01 00:00:01+00:00,view,17302664,2053013553853497655,NaN,creed,28.31,561587266,755422e7-9040-477b-9bd2-6a6e8fd97387
3,2019-11-01 00:00:01+00:00,view,3601530,2053013563810775923,appliances.kitchen.washer,lg,712.87,518085591,3bfb58cd-7892-48cc-8020-2f17e6de6e7f
4,2019-11-01 00:00:01+00:00,view,1004775,2053013555631882655,electronics.smartphone,xiaomi,183.27,558856683,313628f1-68b8-460d-84f6-cec7a8796ef2
...,...,...,...,...,...,...,...,...,...
99995,2019-11-01 03:52:45+00:00,view,1004781,2053013555631882655,electronics.smartphone,huawei,254.71,514148475,03f3fa00-1c0a-48e3-9cd1-f9ae4b11809e
99996,2019-11-01 03:52:45+00:00,view,2600128,2053013563970159485,NaN,NaN,175.55,560904986,344e7ccb-169c-48bc-8b9b-b93bce85391e
99997,2019-11-01 03:52:45+00:00,view,1005159,2053013555631882655,electronics.smartphone,xiaomi,212.10,566309730,72b2106a-c885-41d5-99ab-bbed9fa730f7
99998,2019-11-01 03:52:45+00:00,cart,1801906,2053013554415534427,electronics.video.tv,tcl,257.33,564331418,092b3c00-2d01-40e7-a1d9-dfa9bbad847e


In [90]:
df["user_session"].isna().mean()

np.float64(0.0)

In [91]:
session_size = (
    df.groupby("user_session")
      .size()
)

session_size.describe()

count    24889.000000
mean         4.017839
std          5.465226
min          1.000000
25%          1.000000
50%          2.000000
75%          5.000000
max        140.000000
dtype: float64

In [92]:
sessions_per_user = (
    df.groupby("user_id")["user_session"]
      .nunique()
)

sessions_per_user.describe()

count    20778.000000
mean         1.197998
std          0.742207
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         51.000000
Name: user_session, dtype: float64